src/Eloquent/Casts/ObjectId.php

```diff
@@ -19,28 +27,38 @@ class ObjectId implements CastsAttributes
      */
     public function get($model, string $key, $value, array $attributes)
     {
-        if (! $value instanceof BSONObjectId) {
-            return $value;
+        if ($value instanceof BSONObjectId && $model->getKeyName() === $key && $model->getKeyType() === 'string') {
```

Comment:
**⚠️ IMPORTANT ⚠️**

The `ObjectId` cast should not be used with anything else than `ObjectId` key type. Otherwise there is a mismatch in the type when using `Model->getKey()` to query the database.

This impacts queries using the query builder and relations.

Reply:
This must be revisited to support SQL<->MongoDB relations.